File: rerun-reports/02-python/20260829-032449/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/blob_event_handler.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any
from urllib.parse import unquote

from azure.core.exceptions import HttpResponseError, ResourceModifiedError, ResourceNotFoundError

from event_receiver import ReceivedEvent

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class BlobLocation:
    container: str
    name: str


def parse_blob_subject(subject: str | None) -> BlobLocation:
    if not subject:
        raise ValueError("Blob event is missing its subject")

    container_marker = "/containers/"
    blob_marker = "/blobs/"
    if container_marker not in subject or blob_marker not in subject:
        raise ValueError(f"Unexpected blob event subject: {subject}")

    container_and_blob = subject.split(container_marker, 1)[1]
    container, separator, blob_name = container_and_blob.partition(blob_marker)
    if not separator or not container or not blob_name:
        raise ValueError(f"Unexpected blob event subject: {subject}")
    return BlobLocation(unquote(container), unquote(blob_name))


def _subject(event: ReceivedEvent) -> str | None:
    return event.subject


def _is_expected_race(error: HttpResponseError) -> bool:
    return error.status_code in {404, 409, 412} or getattr(error, "error_code", None) in {
        "BlobArchived",
        "BlobNotFound",
        "ConditionNotMet",
    }


class BlobEventHandler:
    def __init__(self, blob_service_client: Any) -> None:
        self._client = blob_service_client

    def handle_created(self, event: ReceivedEvent) -> None:
        location = parse_blob_subject(_subject(event))
        blob_client = self._client.get_blob_client(location.container, location.name)
        try:
            properties = blob_client.get_blob_properties()
            downloader = blob_client.download_blob()
            downloader.readall()
        except (ResourceNotFoundError, ResourceModifiedError) as error:
            logger.warning(
                "Blob %s/%s changed or disappeared before it could be read: %s",
                location.container,
                location.name,
                error,
            )
            return
        except HttpResponseError as error:
            if not _is_expected_race(error):
                raise
            logger.warning(
                "Blob %s/%s is no longer readable at its original tier: %s",
                location.container,
                location.name,
                error,
            )
            return

        print(
            "Blob created: "
            f"name={location.name}, size={properties.size}, "
            f"content_type={properties.content_settings.content_type}, "
            f"access_tier={properties.blob_tier}"
        )

    def handle_deleted(self, event: ReceivedEvent) -> None:
        location = parse_blob_subject(_subject(event))
        logger.info("Blob deleted: %s/%s", location.container, location.name)


class AsyncBlobEventHandler:
    def __init__(self, blob_service_client: Any) -> None:
        self._client = blob_service_client

    async def handle_created(self, event: ReceivedEvent) -> None:
        location = parse_blob_subject(_subject(event))
        blob_client = self._client.get_blob_client(location.container, location.name)
        try:
            properties = await blob_client.get_blob_properties()
            downloader = await blob_client.download_blob()
            await downloader.readall()
        except (ResourceNotFoundError, ResourceModifiedError) as error:
            logger.warning(
                "Blob %s/%s changed or disappeared before it could be read: %s",
                location.container,
                location.name,
                error,
            )
            return
        except HttpResponseError as error:
            if not _is_expected_race(error):
                raise
            logger.warning(
                "Blob %s/%s is no longer readable at its original tier: %s",
                location.container,
                location.name,
                error,
            )
            return

        print(
            "Blob created (async): "
            f"name={location.name}, size={properties.size}, "
            f"content_type={properties.content_settings.content_type}, "
            f"access_tier={properties.blob_tier}"
        )

    async def handle_deleted(self, event: ReceivedEvent) -> None:
        location = parse_blob_subject(_subject(event))
        logger.info("Blob deleted (async): %s/%s", location.container, location.name)

File: rerun-reports/03-java/20260829-075759/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp/generated-code/src/main/java/com/example/blobevents/BlobEventHandler.java
package com.example.blobevents;

import com.azure.core.util.BinaryData;
import com.azure.storage.blob.BlobAsyncClient;
import com.azure.storage.blob.BlobClient;
import com.azure.storage.blob.BlobServiceAsyncClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.models.BlobProperties;
import com.azure.storage.blob.models.BlobStorageException;
import reactor.core.publisher.Mono;

import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Objects;
import java.util.logging.Logger;

public final class BlobEventHandler {
    private static final Logger LOGGER = Logger.getLogger(BlobEventHandler.class.getName());
    private static final String CONTAINERS_SEGMENT = "/containers/";
    private static final String BLOBS_SEGMENT = "/blobs/";

    private final SyncBlobReader syncReader;
    private final AsyncBlobReader asyncReader;

    public BlobEventHandler(BlobServiceClient syncClient, BlobServiceAsyncClient asyncClient) {
        this(
            location -> readBlob(syncClient, location),
            location -> readBlobAsync(asyncClient, location));
    }

    BlobEventHandler(SyncBlobReader syncReader, AsyncBlobReader asyncReader) {
        this.syncReader = Objects.requireNonNull(syncReader, "syncReader");
        this.asyncReader = Objects.requireNonNull(asyncReader, "asyncReader");
    }

    public void handleCreated(BlobLifecycleEvent event) {
        BlobLocation location = parseSubject(event.subject());
        try {
            logSummary(syncReader.read(location));
        } catch (BlobStorageException exception) {
            if (!handleExpectedRace(location, exception)) {
                throw exception;
            }
        }
    }

    public Mono<Void> handleCreatedAsync(BlobLifecycleEvent event) {
        BlobLocation location = parseSubject(event.subject());
        return asyncReader.read(location)
            .doOnNext(BlobEventHandler::logSummary)
            .then()
            .onErrorResume(BlobStorageException.class, exception -> {
                if (handleExpectedRace(location, exception)) {
                    return Mono.empty();
                }
                return Mono.error(exception);
            });
    }

    public void handleDeleted(BlobLifecycleEvent event) {
        BlobLocation location = parseSubject(event.subject());
        LOGGER.info(() -> "Blob deleted: container=" + location.container() + ", name=" + location.name());
    }

    public Mono<Void> handleDeletedAsync(BlobLifecycleEvent event) {
        handleDeleted(event);
        return Mono.empty();
    }

    static BlobLocation parseSubject(String subject) {
        int containerStart = subject.indexOf(CONTAINERS_SEGMENT);
        int blobSeparator = subject.indexOf(BLOBS_SEGMENT, containerStart + CONTAINERS_SEGMENT.length());
        if (containerStart < 0 || blobSeparator < 0) {
            throw new IllegalArgumentException("Unexpected blob event subject: " + subject);
        }

        String container = subject.substring(containerStart + CONTAINERS_SEGMENT.length(), blobSeparator);
        String name = subject.substring(blobSeparator + BLOBS_SEGMENT.length());
        if (container.isBlank() || name.isBlank()) {
            throw new IllegalArgumentException("Blob event subject must contain a container and blob name: " + subject);
        }
        return new BlobLocation(decodePathPart(container), decodePathPart(name));
    }

    private static BlobSummary readBlob(BlobServiceClient serviceClient, BlobLocation location) {
        BlobClient blob = serviceClient.getBlobContainerClient(location.container()).getBlobClient(location.name());
        blob.downloadContent();
        BlobProperties properties = blob.getProperties();
        return toSummary(location, properties);
    }

    private static Mono<BlobSummary> readBlobAsync(
        BlobServiceAsyncClient serviceClient,
        BlobLocation location
    ) {
        BlobAsyncClient blob = serviceClient.getBlobContainerAsyncClient(location.container())
            .getBlobAsyncClient(location.name());
        return blob.downloadContent()
            .then(blob.getProperties())
            .map(properties -> toSummary(location, properties));
    }

    private static BlobSummary toSummary(BlobLocation location, BlobProperties properties) {
        String accessTier = properties.getAccessTier() == null ? "unknown" : properties.getAccessTier().toString();
        return new BlobSummary(
            location.name(),
            properties.getBlobSize(),
            properties.getContentType(),
            accessTier);
    }

    private static void logSummary(BlobSummary summary) {
        LOGGER.info(() -> "Blob created: name=" + summary.name()
            + ", size=" + summary.size()
            + ", contentType=" + summary.contentType()
            + ", accessTier=" + summary.accessTier());
    }

    private static boolean handleExpectedRace(BlobLocation location, BlobStorageException exception) {
        String errorCode = exception.getErrorCode() == null ? "" : exception.getErrorCode().toString();
        if (exception.getStatusCode() == 404) {
            LOGGER.warning(() -> "Blob disappeared before it could be read: " + location);
            return true;
        }
        if (exception.getStatusCode() == 409
            && (errorCode.contains("Archive") || errorCode.contains("Rehydr"))) {
            LOGGER.warning(() -> "Blob is temporarily unreadable in its current access tier: " + location
                + " (" + errorCode + ")");
            return true;
        }
        return false;
    }

    private static String decodePathPart(String value) {
        return URLDecoder.decode(value.replace("+", "%2B"), StandardCharsets.UTF_8);
    }

    record BlobLocation(String container, String name) {
    }

    record BlobSummary(String name, long size, String contentType, String accessTier) {
    }

    @FunctionalInterface
    interface SyncBlobReader {
        BlobSummary read(BlobLocation location);
    }

    @FunctionalInterface
    interface AsyncBlobReader {
        Mono<BlobSummary> read(BlobLocation location);
    }
}
